Distribution/Server/Util/Markdown.hs

```diff
@@ -0,0 +1,121 @@
+  strong ils = HHtml (strong $ unHHtml ils)
+  link target title ils = HHtml $
+    link (adjustRelativeLink target) title (unHHtml ils)
+  image target title ils = HHtml $
+    image (adjustRelativeLink target) title (unHHtml ils)
+  code t = HHtml $ code t
+  rawInline f t = HHtml (rawInline f t)
+
+instance (Rangeable (Html a), IsInline (HHtml a))
+     => IsBlock (HHtml a) (HHtml a) where
+  paragraph ils = HHtml $ paragraph $ unHHtml ils
+  plain ils = HHtml $ plain $ unHHtml ils
+  thematicBreak = HHtml thematicBreak
+  blockQuote bs = HHtml $ blockQuote $ unHHtml bs
+  codeBlock info t = HHtml $ codeBlock info t
+  heading level ils = HHtml $ heading level $ unHHtml ils
+  rawBlock f t = HHtml $ rawBlock f t
+  referenceLinkDefinition x y = HHtml $ referenceLinkDefinition x y
+  list lType lSpacing items =
+    HHtml $ list lType lSpacing $ map unHHtml items
+
+instance HasEmoji (HHtml a) where
+  emoji kw cs = HHtml $ emoji kw cs
+
+instance HasStrikethrough (HHtml a) where
+  strikethrough ils = HHtml $ strikethrough $ unHHtml ils
+
+instance HasPipeTable (HHtml a) (HHtml a) where
+  pipeTable aligns heads rows =
+    HHtml $ pipeTable aligns (map unHHtml heads) (map (map unHHtml) rows)
+
+instance (Rangeable (Html a), Rangeable (HHtml a))
+         => HasTaskList (HHtml a) (HHtml a) where
+  taskList ltype lspacing items =
+    HHtml $ taskList ltype lspacing
+          $ map (\(done, bl) -> (done, unHHtml bl)) items
+
+instance HasMath (HHtml a) where
+  inlineMath t = HHtml $ inlineMath t
+  displayMath t = HHtml $ displayMath t
+
+
+instance Rangeable (Html a) => HasFootnote (HHtml a) (HHtml a) where
+  footnote x y (HHtml t) = HHtml (footnote x y t)
+  footnoteList xs = HHtml $ footnoteList (map unHHtml xs)
+  footnoteRef x y (HHtml t) = HHtml (footnoteRef x y t)
+
+adjustRelativeLink :: T.Text -> T.Text
+adjustRelativeLink url
+  | isRelativeReference (T.unpack url) &&
+    not ("/" `T.isPrefixOf` url)
+              = "src/" <> url
+  | otherwise = url
+
+
+renderHHtml :: HHtml () -> TL.Text
+renderHHtml (HHtml x) = renderHtml x
+
+renderMarkdown :: String -> BS.ByteString -> XHtml.Html
+renderMarkdown name md =
+     either (const $ XHtml.pre XHtml.<< T.unpack txt) (XHtml.primHtml . T.unpack . sanitizeBalance . TL.toStrict . (renderHtml :: Html () -> TL.Text)) $
+         runIdentity (commonmarkWith (mathSpec <> footnoteSpec <> defaultSyntaxSpec <> gfmExtensions)
+                     name
+                     txt)
+  where txt = T.decodeUtf8With T.lenientDecode . BS.toStrict $ md
+
+renderMarkdownRel :: String -> BS.ByteString -> XHtml.Html
+renderMarkdownRel name md =
+     either (const $ XHtml.pre XHtml.<< T.unpack txt) (XHtml.primHtml . T.unpack . sanitizeBalance . TL.toStrict . renderHHtml) $
+         runIdentity (commonmarkWith (mathSpec <> footnoteSpec <> defaultSyntaxSpec <> gfmExtensions)
+                     name
+                     txt)
+  where txt = T.decodeUtf8With T.lenientDecode . BS.toStrict $ md
```

Comment:
When I was a kid, there were these puzzles "Can you see the difference?" with two identically looking pictures next to each other where you had to stare for minutes to see the stroke missing on one side and another stroke missing on the other side.
Hint: the diff between `renderMarkdown` and `renderMarkdownRel` is one letter (at least that is how far I got from staring for minutes).

Reply:
Indeed. One uses HHtml and the other Html. The former renders some links with relative paths to the src dir. A refactor and comments would be welcome.

Reply:
See #1003.